design/BufferOperations.md

```diff
@@ -0,0 +1,150 @@
+# Buffer operations
+
+This describes operations that are done on `WebGPUBuffer` objects directly and are not buffered inside a `WebGPUCommandBuffer`.
+The two primitives we need to support are the CPU writing data inside the buffer for use by the GPU (upload) and the CPU reading data produced by the GPU (readback).
+
+Design constraints are:
+
+ - For the portability of the API, prevent data races between the CPU and the GPU.
+ - For performance, minimize the number of times the data is copied around.
+ - To make the API non-blocking, only allow asynchronous readbacks.
+ - For performance on multi-process implementations, make an asynchronous upload path.
+
+## Buffer mapping
+
+### map[Write|Read] and unmap
+
+The way to have the minimal number of copies for upload and readback is to provide a buffer mapping mechanism.
+This mechanism has to be asynchronous to ensure the GPU is done using the buffer before the application can look into the ArrayBuffer.
+Otherwise on implementation where the ArrayBuffer is directly a pointer to the buffer memory, data races between the CPU and the GPU could occur.
+
+We want the status of a map operation to act as both a promise, and something that's pollable as there are advantages to both.
+`WebGPUMappedMemory` is an object that is `then`-able, meaning that it acts like a Javascript `Promise` but is pollable at the same time.
+
+The mapping operations for `WebGPUBuffer` are:
+
+```
+partial interface WebGPUBuffer {
+    WebGPUMappedMemory mapWrite(u32 offset, u32 size);
+    WebGPUMappedMemory mapRead(u32 offset, u32 size);
+};
+```
+
+These operations return new `WebGPUMappedMemory` objects representing the current range of the buffer for writing or mapping.
+The results are initialized in the "pending" state and transition at Javascript task boundary to the "available" state when the implementation can determine the GPU is done using the buffer.
```

Comment:
For this or a subsequent pull request, I suggest making a state diagram that illustrates all of the states a `WebGPUBuffer` can be in (available, pending, mapped, unmapped) along with the operations that move it between states (calling unmap, calling map, GPU writes finished, JavaScript task boundary, etc)

Reply:
See the state diagrams in this [doc](https://docs.google.com/document/d/1HFzMMvDGHFtTgjNT0j-0SQ1fNU9R7woZ4JuNJdAXBjg/edit#heading=h.67ockbzesm2a). I could include them here but they assume D3D12 style memory barriers.